src/main/java/soot/jbco/jimpleTransformations/MethodRenamer.java

```diff
@@ -214,51 +214,52 @@ public static String getNewName() {
+	private static boolean allowsRename(SootClass sc, SootMethod method) {
+		if (soot.jbco.Main.getWeight(MethodRenamer.name, method.getName()) == 0) {
 			return false;
+		}
 
-		String subSig = m.getSubSignature();
-		if (subSig.equals("void main(java.lang.String[])") && m.isPublic()
-				&& m.isStatic()) {
+		String subSig = method.getSubSignature();
+		if ("void main(java.lang.String[])".equals(subSig) && method.isPublic() && method.isStatic()) {
 			return false; // skip the main method - it needs to be named 'main'
-		} else if (subSig.indexOf("void <init>(") >= 0
-				|| subSig.equals("void <clinit>()")) {
+		} else if (subSig.contains(SootMethod.constructorName) || subSig.equals(SootMethod.staticInitializerName)) {
 			return false; // skip constructors for now
 		} else {
-			for (SootClass _c : hierarchy.getSuperclassesOfIncluding(c
-					.getSuperclass())) {
-				if (_c.isLibraryClass() && _c.declaresMethod(subSig)
-						&& hierarchy.isVisible(c, _c.getMethod(subSig))) {
-					return false;
-				}
-			}
-
-			do {
-				if (checkInterfacesForMethod(c, m))
-					return false;
-			} while (c.hasSuperclass() && (c = c.getSuperclass()) != null);
+			return !(isOverriddenLibraryInterfaceMethod(sc, method) || isOverriddenLibrarySuperclassMethod(sc, method));
 		}
-
-		return true;
 	}
 
-	private static boolean checkInterfacesForMethod(SootClass c, SootMethod m) {
-		for (SootClass sc : c.getInterfaces()) {
-			if (sc.isLibraryClass()
-					&& sc.declaresMethod(m.getName(), m.getParameterTypes(),
-							m.getReturnType()))
-				return true;
-		}
-		return false;
-	}
+    private static boolean isOverriddenLibrarySuperclassMethod(SootClass sc, SootMethod method) {
+        String subSignature = method.getSubSignature();
+        //use getSuperclassesOfIncluding instead of getSuperclassesOf to avoid problems when 'sc' parameter is interface
+        return hierarchy.getSuperclassesOfIncluding(sc.getSuperclass()).stream()
+                .filter(SootClass::isLibraryClass)
+                .filter(c -> c.declaresMethod(subSignature))
+                .anyMatch(c -> hierarchy.isVisible(sc, c.getMethod(subSignature)));
+    }
+
+    private static boolean isOverriddenLibraryInterfaceMethod(SootClass sc, SootMethod method) {
+        return getAllInterfacesOf(sc).stream()
+                .filter(SootClass::isLibraryClass)
+                .anyMatch(c -> c.declaresMethod(method.getName(), method.getParameterTypes(), method.getReturnType()));
+    }
+
+    private static List<SootClass> getAllInterfacesOf(SootClass sc) {
+        Stream<SootClass> superClassInterfaces = sc.isInterface() ? Stream.empty() : hierarchy.getSuperclassesOf(sc)
+                .stream()
+                .map(MethodRenamer::getAllInterfacesOf)
+                .flatMap(Collection::stream);
+        Stream<SootClass> directInterfaces = Stream.concat(sc.getInterfaces().stream(), sc.getInterfaces().stream()
```

Comment:
You concat the same stream here, is this correct? If so, could you please explain the semantics behind it?

Reply:
Please re-check this, second stream is not the same as first.